### PR TITLE
Adds "notes per second" limits for alternate and single-tap mods

### PR DESCRIPTION
**Pull request by o-dasher**
This pull request aims to bring more customization into alternate and singletap mods.

Sometimes you don't want to be forced to alternate some kind of level of note density or either being forced to single tap some kind of note density too, that's why i am bringing this changes.

I remember someone wrote an issue requesting for this feature (I am not sure though). although it was being requested as a "minimal bpm" setting, i believe notes per second should fit this role better and is more accessible for the average user, but i can change to bpm if required.

Pull request changes
- Customize the minimal number of notes per second that you want to begin being forced to alternate with the alternate mod. 
- Customize the number of notes per second that you want to not be required to singletap with the singletap mod.

### CODE REVIEW

**Comment by Walavouchey** [comment]
i haven't tested gameplay with this but from the concept i'm concerned it'll be hard to tell when you're allowed to singletap/alternate and when you don't need to. notes per second may vary drastically in a short time and throughout a map, so this would ideally need to be telegraphed in some way

**Comment by Moltanicaa** [comment]
> i haven't tested gameplay with this but from the concept i'm concerned it'll be hard to tell when you're allowed to singletap/alternate and when you don't need to.

Maybe different colour of combo? It seems to me that if for ordinary singletap moments it should be, for example, cold colors, then for acceptable alternative moments it should be warm color.

**Comment by o-dasher** [comment]
> Maybe different colour of combo? It seems to me that if for ordinary singletap moments it should be, for example, cold colors, then for acceptable alternative moments it should be warm color.
I think this isn't that hard to implement, and seems like a great way to go around with this problem, once i get home i will try doing this.

File: osu.Game.Rulesets.Osu/Mods/InputBlockingMod.cs [hunk 1]
```diff
@@ -110,5 +126,14 @@ public void OnReleased(KeyBindingReleaseEvent<OsuAction> e)
             {
             }
         }
+
+        protected class NotesPerSecondSetting : BindableInt
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/InputBlockingMod.cs:130]: Why is this its own class?

**Comment by peppy** [comment]
I dunno about this, seems like potential settings clutter. And very hard to be useful without visuals. How is a user supposed to even know what to set this to.

**Comment by o-dasher** [comment]
Maybe i should go with Beats Per Minute based, since a user can set it to a bpm the user is comfortable alternating with / or want to practice alternating to.

File: osu.Game.Rulesets.Osu/Mods/InputBlockingMod.cs (hunk 1 above)
Comment (o-dasher) [reply osu.Game.Rulesets.Osu/Mods/InputBlockingMod.cs:130]: It's just to avoid code repetition between the two options.

**Comment by peppy** [comment]
Feels like if that's the purpose, maybe the mod should just be adjusted to appease the majority?

Or just a toggle being like "don't require alternating on slow sections".

**Comment by o-dasher** [comment]
> Feels like if that's the purpose, maybe the mod should just be adjusted to appease the majority?
> 
> Or just a toggle being like "don't require alternating on slow sections".

I didn't understand what you mean by the first sentence.

For the other one, yeah this sounds nice but what could be defined as a slow section though?

**Comment by peppy** [comment]
> > Feels like if that's the purpose, maybe the mod should just be adjusted to appease the majority?
> > Or just a toggle being like "don't require alternating on slow sections".
> 
> I didn't understand what you mean by the first sentence.

I mean just adjusting the threshold requirement on the mod, with no setting.

> For the other one, yeah this sounds nice but what could be defined as a slow section though?

I dunno, poll users or take a sane guess and adjust from there. You could probably check autoplay's algorithm to see where it starts alternating.

**Comment by smoogipoo** [comment]
Thanks for the contribution, but I'm going to take a stance and say that this is too much customisation for the time being.

**Comment by Xotna** [comment]
Just a singletap user point of view here: i was looking forward to such a change since training high bpm singletap with singletap mod quickly becomes a chore whenever the map has 1/4 bursts. I hope this will be reconsidered in the future, and sorry for the noise.

**Comment by peppy** [comment]
To reiterate, I don't think we're against changing how the mod behaves (ie. better defaults), but it needs more discussion before adding customisation overload. We'd want to provide both the best default experience, and if that can't serve all, the ability to customise in the simplest way possible (not manually figuring out the precise notes / beats per second required).

**Pull request closed without merging**